src/encryption_layer.rs

```diff
@@ -0,0 +1,248 @@
+use std::{fmt::Debug, ops::Deref, sync::Arc};
+
+use crate::{
+    ADDRESS_LENGTH, KEY_LENGTH, MemoryADT, address::Address, error::Error, secret::Secret,
+    symmetric_key::SymmetricKey,
+};
+use aes::{
+    Aes256,
+    cipher::{BlockEncrypt, KeyInit, generic_array::GenericArray},
+};
+use xts_mode::Xts128;
+
+#[derive(Clone)]
+struct ClonableXts(Arc<Xts128<Aes256>>);
+
+impl Debug for ClonableXts {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("ClonableXts").finish()
+    }
+}
+
+impl Deref for ClonableXts {
+    type Target = Xts128<Aes256>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// The encryption layers is built on top of an encrypted memory implementing the `MemoryADT` and
+/// exposes a plaintext virtual memory interface implementing the `MemoryADT`.
+///
+/// This type is thread-safe.
+#[derive(Debug, Clone)]
+pub struct MemoryEncryptionLayer<
+    const WORD_LENGTH: usize,
+    Memory: MemoryADT<Address = Address<ADDRESS_LENGTH>>,
+> {
+    aes: Aes256,
+    xts: ClonableXts,
+    mem: Memory,
+}
+
+impl<
+    const WORD_LENGTH: usize,
+    Memory: Send + Sync + MemoryADT<Address = Address<ADDRESS_LENGTH>, Word = [u8; WORD_LENGTH]>,
+> MemoryEncryptionLayer<WORD_LENGTH, Memory>
+{
+    /// Instantiates a new memory encryption layer.
+    pub fn new(seed: Secret<KEY_LENGTH>, stm: Memory) -> Self {
+        let k_p = SymmetricKey::<KEY_LENGTH>::derive(&seed, &[0]).expect("secret is large enough");
+        let k_e1 =
+            SymmetricKey::<{ KEY_LENGTH }>::derive(&seed, &[1]).expect("secret is large enough");
+        let k_e2 =
+            SymmetricKey::<{ KEY_LENGTH }>::derive(&seed, &[2]).expect("secret is large enough");
+        let aes = Aes256::new(GenericArray::from_slice(&k_p));
+        let aes_e1 = Aes256::new(GenericArray::from_slice(&k_e1));
+        let aes_e2 = Aes256::new(GenericArray::from_slice(&k_e2));
+        // The 128 in the XTS name refer to the block size. AES-256 is used here, which confers
+        // 128 bits of PQ security.
+        let xts = ClonableXts(Arc::new(Xts128::new(aes_e1, aes_e2)));
+        Self { aes, xts, mem: stm }
+    }
+
+    /// Permutes the given memory address.
+    fn permute(&self, mut a: Memory::Address) -> Memory::Address {
```

Comment:
```suggestion
    fn permute(&self, mut addr: Memory::Address) -> Memory::Address {
```

Reply:
I prefer to stick with the following naming convention:
- `a` for address
- `w` for word
- `v` for value

These have the advantage of being single-letter (easy to read once you get them, do not waste space), rather intuitive, and types explicit full letter names anyway.

Reply:
OK